_posts/2020-12-25-manual-gates-git-flow-in-a-wig.md

```diff
@@ -0,0 +1,145 @@
+
+1. A simple pipeline is constructed that deploys straight to production
+
+<center><img src="/img/manual-gates/evolve-1.png" /></center>
+<br/>
+
+2. A pre-production environment is created with a gate, in order to manually verify an environment before it reaches production.
+
+<center><img src="/img/manual-gates/evolve-2.png" /></center>
+<br/>
+
+3. The tests that are run during gating are automated, there-by replacing the automated gate with automatic verification.
+
+<center><img src="/img/manual-gates/evolve-3.png" /></center>
+<br/>
+
+4. The automated test suite begins to get slower as it both increases in size and complexity. The developers notice that this is due to bootstrapping external resources. The developers create an additional pre-production environment, and split the tests. Tests are run with mocked dependencies against the first pre-production environment, where-as real dependencies are used against the second pre-production environment.
+
+<center><img src="/img/manual-gates/evolve-4.png" /></center>
+<br/>
+
+Each of these environments is about stopping faulty or incorrect software hitting production. Concerns such as User Acceptance Testing should not block the path to production - this should have happened before code hit the master branch. Training environments should not be poached from the environments in the main pipeline. If I haven't been clear up to this point, let me restate it - the environments that are present in your main pipeline to production should not have any other function but ensuring that you can reliably deploy code to production that will not cause an unintentional degredation.
+
+To continue that point, I often seem teams that will have a "development" environment in their pipeline. This is a step above developers running entire copies of the system in their local environment, but it still comes with issues. It can be difficult to accurately determine exactly what is running in this environment when multiple developers are deploying changes to such an environment, which can make it difficult to ensure working code is making it's way towards production. This is often ensures that a manual gate is introduced and that it never leaves. If developers must be able to run an isolated copy of the system, it is better to implement this via developer sandboxes or dynamic environments that are free of the pipeline. Recognise that the purpose of such an environment isn't to safely deploy code to production, but to aide the developer in feature development, and as such it doesn't belong in the production pipeline.
+
+# A Flexible Model For Software Delivery Pipelines
+
+Assume we are deploying software-as-a-service. This could be an internal or customer facing service.
```

Comment:
Up to this point, your arguments are clear and concise:
* gitflow bad
* manual gates bad
* non-supporting environments bad

However, in this section you go off in a tangent to discuss alternative pipeline approaches.  Feels like it undermines your points above.

Would this be better as a separate follow up post?  This post you've introduced what your delivery pipeline should, and should not, be.  As a follow up post, you can explore some of the alternate requirements and how you can satisfy them in more detail.

Reply:
I think what I was trying to get is that you have one 'path-to-production' - your master/main pipeline, and that needs to be all about making sure the code gets into production as quick as possible. Trying to stick additional testing, training, and development stages undermines that - so make sure releases to those environments (if required) are completely decoupled from the master pipeline.

I thought I might have needed a concrete example as to what I mean by this?

Reply:
Yeah it's an interesting point/take.  

Personally, I feel that there's underlying waste that is covering those assumptions (why shouldn't people train in production, etc),but you're right, these concerns are orthogonal to your value stream.

I think there's enough here to explore in a separate post (with examples, etc): 
* This post is about your value stream and the path to production.
* This section is about people getting the value stream confused with deployment streams (new term :shrug: ?).  Which is an extension/follow-up to your value stream argument - with a separate conclusion

Reply:
You might have legitimate reasons - e.g. If the system you are training people is on something like a train control system or something. I suppose maybe it's less about 'production' or more about 'tenancy'. If the production app has a concept of data isolation it is perhaps less of an issue (but so many don't, and it may be better from a cost perspective to clone the environment then build in multi-tenancy at the application level).

So maybe there is a better post here around data isolation / tenancy being a different concern then the value stream. Maybe you achieve that duplicated infrastructure, and maybe you don't.